File: backend/models/location.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// LocationType defines the type of location
type LocationType string

const (
	LocationTypeGudang LocationType = "gudang"
	LocationTypeToko   LocationType = "toko"
)

// Location represents warehouse or store location
type Location struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	Code        string         `gorm:"not null;size:20" json:"code"` // unique index created manually in migration
	Name        string         `gorm:"not null;size:100" json:"name"`
	Type        LocationType   `gorm:"not null;size:20;index" json:"type"` // gudang or toko
	Address     string         `gorm:"size:255" json:"address"`
	Phone       string         `gorm:"size:20" json:"phone"`
	Description string         `gorm:"size:255" json:"description"`
	IsActive    bool           `gorm:"default:true" json:"is_active"`
	Boxes       []StorageBox   `gorm:"foreignKey:LocationID" json:"boxes,omitempty"`
	Stocks      []Stock        `gorm:"foreignKey:LocationID" json:"stocks,omitempty"`
}

// StorageBox represents a storage box within a location
type StorageBox struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	LocationID  uint           `gorm:"not null;index" json:"location_id"`
	Location    Location       `gorm:"foreignKey:LocationID" json:"location,omitempty"`
	Code        string         `gorm:"not null;size:20;index" json:"code"` // e.g., "A1", "B2", etc.
	Name        string         `gorm:"not null;size:50" json:"name"`       // e.g., "Kotak A1"
	Description string         `gorm:"size:255" json:"description"`
	Capacity    int            `gorm:"default:0" json:"capacity"` // Max items this box can hold (0 = unlimited)
	IsActive    bool           `gorm:"default:true" json:"is_active"`
	Stocks      []Stock        `gorm:"foreignKey:StorageBoxID" json:"stocks,omitempty"`
}

// UniqueIndex for box code within a location
func (StorageBox) TableName() string {
	return "storage_boxes"
}

File: mobile/lib/models/location.dart
import 'package:equatable/equatable.dart';
import 'storage_box.dart';
import 'stock.dart';

class Location extends Equatable {
  final int id;
  final String code;
  final String name;
  final String type; // gudang, toko
  final String address;
  final String phone;
  final String description;
  final bool isActive;
  final List<StorageBox>? boxes;
  final List<Stock>? stocks;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Location({
    required this.id,
    required this.code,
    required this.name,
    required this.type,
    this.address = '',
    this.phone = '',
    this.description = '',
    this.isActive = true,
    this.boxes,
    this.stocks,
    this.createdAt,
    this.updatedAt,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      id: json['id'] ?? 0,
      code: json['code'] ?? '',
      name: json['name'] ?? '',
      type: json['type'] ?? 'toko',
      address: json['address'] ?? '',
      phone: json['phone'] ?? '',
      description: json['description'] ?? '',
      isActive: json['is_active'] ?? true,
      boxes: json['boxes'] != null
          ? (json['boxes'] as List).map((b) => StorageBox.fromJson(b)).toList()
          : null,
      stocks: json['stocks'] != null
          ? (json['stocks'] as List).map((s) => Stock.fromJson(s)).toList()
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'name': name,
      'type': type,
      'address': address,
      'phone': phone,
      'description': description,
      'is_active': isActive,
    };
  }

  Location copyWith({
    int? id,
    String? code,
    String? name,
    String? type,
    String? address,
    String? phone,
    String? description,
    bool? isActive,
    List<StorageBox>? boxes,
    List<Stock>? stocks,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Location(
      id: id ?? this.id,
      code: code ?? this.code,
      name: name ?? this.name,
      type: type ?? this.type,
      address: address ?? this.address,
      phone: phone ?? this.phone,
      description: description ?? this.description,
      isActive: isActive ?? this.isActive,
      boxes: boxes ?? this.boxes,
      stocks: stocks ?? this.stocks,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [id, code, name, type, isActive];
}
